src/Nethermind/Nethermind.Consensus/Processing/CensorshipDetector.cs

```diff
@@ -0,0 +1,91 @@
+// SPDX-FileCopyrightText: 2024 Demerzel Solutions Limited
+// SPDX-License-Identifier: LGPL-3.0-only
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using Nethermind.Blockchain;
+using Nethermind.Core;
+using Nethermind.Core.Caching;
+using Nethermind.Int256;
+using Nethermind.TxPool;
+
+namespace Nethermind.Consensus.Processing;
+
+public class CensorshipDetector : IDisposable
+{
+    private readonly ITxPool _txPool;
+    private readonly IBlockProcessor _blockProcessor;
+    private readonly IBlockTree _blockTree;
+    private readonly IAccountStateProvider _accounts;
+
+    private const int CacheSize = 4;
+    private readonly LruCache<long, bool>
+        _censorshipDetector = new(CacheSize, CacheSize, "censorshipDetector");
+
+    public CensorshipDetector(
+        ITxPool txPool,
+        IBlockProcessor blockProcessor,
+        IBlockTree blockTree,
+        IAccountStateProvider accounts)
+    {
+        _txPool = txPool;
+        _blockProcessor = blockProcessor;
+        _blockTree = blockTree;
+        _accounts = accounts;
+        _blockProcessor.BlockProcessing += OnBlockProcessing;
+        _blockTree.NewHeadBlock += OnNewHead;
+    }
+
+    private void OnBlockProcessing(object? sender, BlockEventArgs e)
+    {
+        Task.Run(() => Cache(e.Block));
+    }
+
+    private void OnNewHead(object? sender, BlockEventArgs e)
+    {
+        Task.Run(() => Cache(e.Block));
+    }
```

Comment:
Now you can easily cache same block twice, is this desired?

Reply:
No, this is a mistake. We have to distinguish between what we do when block is processed and when block is added. Potential solution is to maintain an additional slightly bigger cache for processed blocks. As the blocks get confirmed, they are simply added to the main cache.